From b712fafa2e19671e7466e3cfb6a30ae29616de4c Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 14 Sep 2026 15:25:42 +0800
Subject: [PATCH] mtk: mt76: mt7915: Fix per band token limitation

Fix incorrect token decrement logic in mt76_connac2_tx_token_put
function by ensuring the phy token count is decremented alongside the
device token count. Additionally, initialize the token threshold based
on DBDC support during device registration to prevent potential
misconfigurations.

Feature:
Stability

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt76_connac_mac.c | 3 +++
 mt7915/init.c     | 1 +
 2 files changed, 4 insertions(+)

diff --git a/mt76_connac_mac.c b/mt76_connac_mac.c
index b651ec1df..02b39b028 100644
--- a/mt76_connac_mac.c
+++ b/mt76_connac_mac.c
@@ -1225,8 +1225,11 @@ void mt76_connac2_tx_token_put(struct mt76_dev *dev)
 
 	spin_lock_bh(&dev->token_lock);
 	idr_for_each_entry(&dev->token, txwi, id) {
+		struct mt76_phy *phy = mt76_dev_phy(dev, txwi->phy_idx);
+
 		mt76_connac2_txwi_free(dev, txwi, NULL, NULL);
 		dev->token_count--;
+		phy->tokens--;
 	}
 	spin_unlock_bh(&dev->token_lock);
 	idr_destroy(&dev->token);
diff --git a/mt7915/init.c b/mt7915/init.c
index 4300d66cf..0b1590045 100644
--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -1356,6 +1356,7 @@ int mt7915_register_device(struct mt7915_dev *dev)
 
 	ieee80211_queue_work(mt76_hw(dev), &dev->init_work);
 
+	dev->mt76.token_threshold = dev->mt76.token_size / (1 + !!dev->dbdc_support);
 	dev->recovery.hw_init_done = true;
 
 	ret = mt7915_init_debugfs(&dev->phy);
-- 
2.45.2

